Runtime engine

cclerget opened the pull request:
**Description of the Pull Request (PR):**

- Runtime engine
- Configuration file parser
- Start OCI configuration API for CLI and runtime

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [ ] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [ ] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

cclerget committed:
Runtime engine.
Start OCI configuration API.
Configuration file parser.

cclerget committed:
Merge remote-tracking branch 'upstream/development-3.0' into development-3.0-runtime

cclerget committed:
gofmt ...

bauerm97 commented on pkg/util/capabilities/capabilities.go:
```diff
@@ -0,0 +1,367 @@
+package capabilities
```
Thanks for doing this Cedric, this was something I needed to get around to implementing but hadn't had time yet :-)

bauerm97 commented on runtime/run.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+echo "suid"
+./cli suid /tmp/testing.simg /bin/ls /
```
What work is left to be done to run SIF images here? I assume we need to use SIF binary to expose/mount the underlying FS?

cclerget replied:
Yes, Yannick could confirm, but I believe we need sif_load in core/lib/sif/sif.c to read SIF image and get data object descriptors to identify partitions and their offsets to mount correct ones

bauerm97 commented on pkg/runtime/oci/config/solaris.go:
```diff
@@ -0,0 +1,3 @@
+// +build solaris
```
Do we need to include Solaris and Windows files? We're not planning on supporting containers on these systems, strictly Linux. Is it necessary for other reasons to have this here?

cclerget replied:
Indeed, it's not necessary, I just split OCI spec between platform to make things clearer and easier

bauerm97 commented on pkg/runtime/oci/config/config_linux.go:
```diff
@@ -0,0 +1,9 @@
+
+const Platform = "linux"
+
+type RuntimeOciPlatform struct {
```
I'm slightly confused how you're using the OCI runtime config. I see you have defined an interface, does the implementation of the interface methods happen in the OCI runtime spec package?

cclerget replied:
No it just describe the whole interface with methods to implement in order to manipulate OCI specs

bauerm97 commented on pkg/runtime/oci/config/root.go:
```diff
@@ -0,0 +1,50 @@
+	SetReadOnly(enabled bool)
+}
+
+type DefaultRuntimeOciRoot struct {
```
Why are each of these `DefaultRuntimeOci*` structs embedding a `RuntimeOciSpec` directly instead of the proper member. For instance, here why are we not having `RuntimeOciSpecRoot *RuntimeOciSpecRoot`?

cclerget replied:
I take the easy way there, because if you embed the proper member, somewhere you have to link it against a common specs.Spec instance to generate JSON at the end

bauerm97 replied:
Ahhh, makes sense. Thanks for clarifying :)

bauerm97 commented on pkg/runtime/runtime.go:
```diff
@@ -1,13 +1,48 @@
+}
+
+// Runtime operations
+type Runtime interface {
```
This would be a good place to start the investigation into runtime plugins.

bauerm97 commented on pkg/runtime/runtime.go:
```diff
@@ -1,13 +1,48 @@
+	Runtime
+}
+
+type RuntimeCLI struct {
```
Is this supposed to be `type RuntimeOCI struct` by any chance?

cclerget replied:
Yes, it embed OciRuntime interface. As discussed yesterday OciRuntime interface need to be on both side

bauerm97 replied:
@cclerget I think you have a typo then ;) This reads `RuntimeCLI` instead of `RuntimeOCI`

cclerget replied:
@bauerm97 not a typo, RuntimeCLI struct contains OciRuntime interface for CLI side. I need to do the same RuntimeEngine

bauerm97 commented on internal/pkg/runtime/runtime.go:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+func init() {
```
I think each `RuntimeEngine` should have an init in its own `.go` file. For example, the OCI transports package has an example of how to pull off this registerEngine effect. See:

https://github.com/containers/image/blob/master/transports/transports.go For how they describe the registering of image transports
https://github.com/containers/image/blob/master/transports/alltransports/alltransports.go For a list of examples of image transports that implement this

cclerget replied:
I took this way because my first thought was to embed all engines in one binary, so to not import all engines each time, I wrote a single init which register all engines resulting in one import for all engines. If we split engines into binaries, an init by engine is the right thing to do

bauerm97 reviewed (changes requested):
Hey @cclerget it's looking really good so far. I haven't had time to pull it locally and play with it yet so I don't want to approve until I do that. 

That being said, I really like the direction this is going and I think we're on the right path!

cclerget committed:
Remove runtime/cli binary from previous commit.
Add missing config.go.
Replace socketpair by pipes.
Add runtime/setup.sh into compile.sh.

cclerget committed:
Modify some parts in core/lib.
Exec/Shell/Run actions call wrapper support namespaces and debug/verbose flags.
Runtime mount for ext3/squashfs/sif/directory images.

bauerm97 reviewed (approved):
Looks good! A few clean up things to fix and I can fix that bug you've described, but I'll get those into a next PR. Awesome work
